### PR TITLE
gh-126316: Make grp.getgrall() thread-safe: add a mutex

### PR DESCRIPTION
**Pull request by vstinner**
grpmodule.c is no longer built with the limited C API, since PyMutex is excluded from the limited C API.

* Issue: gh-126316

### CODE REVIEW

**Comment by vstinner** [comment]
@ZeroIntensity @colesbury: Would you mind to review my change?

**Review by ZeroIntensity — COMMENTED** [review]
Hmm, this is an interesting way to do it. Local `static` variables are notorious for being non-thread-safe, so I'm not sure this kind of approach will work. If it does, I don't think we're fixing any issues with re-entrancy.

I would just go with `@critical_section` on each function, and then switch the module's `Py_mod_multiple_interpreters` to `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` instead of `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED` to denote shared-GIL for subinterpreters. That should fix races for both cases.

**Review by colesbury — APPROVED** [review]
This approach looks good to me. I think you are right that we don't have to worry about deadlock here because the calls inside the lock won't call arbitrary Python code.

There's a failing CI related to generated files. I think the `pycore_lock.h` include in  `pycore_modsupport.h` can be removed now. Also, maybe the arg clinic files need to be regenerated?

> Local static variables are notorious for being non-thread-safe

A static `PyMutex` is fine because `PyMutex` is thread-safe. (It doesn't matter if it's a local or global static variable.)

The general problems with static variables (including static local variables) is that the data structures are not thread-safe or that they are lazily initialized pointers, which is not the case here.

**Comment by ZeroIntensity** [comment]
Thanks for the insight, Sam. How scalable is this approach (as in, will we have a static `PyMutex` everywhere)?

**Comment by colesbury** [comment]
I don't think we'll have `static PyMutex` everywhere, but this is a good pattern for certain libc functions that are not thread-safe and do not have thread-safe alternatives.

**Comment by vstinner** [comment]
> I think the pycore_lock.h include in pycore_modsupport.h can be removed now.

Correct: I wrote https://github.com/python/cpython/pull/127093 to remove it.

File: Modules/grpmodule.c [hunk 1]
```diff
@@ -281,23 +280,33 @@ static PyObject *
...
-    PyObject *d;
-    struct group *p;
-
-    if ((d = PyList_New(0)) == NULL)
+    PyObject *d = PyList_New(0);
+    if (d == NULL) {
         return NULL;
+    }
+
+    static PyMutex getgrall_mutex = {0};
```
Comment (ZeroIntensity) [inline Modules/grpmodule.c:288]: I'm not sure if `PyMutex` likes being const, but this should fix the C analyzer.
```suggestion
    static const PyMutex getgrall_mutex = {0};
```

Comment (colesbury) [reply Modules/grpmodule.c:288]: No, it should not be const.

Comment (ZeroIntensity) [reply Modules/grpmodule.c:288]: Ok, we'll need to do something about the analyzer then if your plan is to use this pattern for other thread-unsafe libc functions. I can probably submit a PR for that sometime today. @vstinner, in the meantime, add this to `ignored.tsv` to fix CI.

Comment (vstinner) [reply Modules/grpmodule.c:288]: > @vstinner, in the meantime, add this to ignored.tsv to fix CI.

Sure, done.

**Review by ZeroIntensity — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
Merged, thanks for reviews @colesbury and @ZeroIntensity.

**Pull request merged**